### ISSUE TITLE
Switch to using 0-indexed spans

### ISSUE THREAD

**Issue by ReeceHumphreys** [icerpc/slicec#669]
Currently, our spans are 1-indexed to be more in line with how files are represented in editors. However, this has proven to be painful when implementing things like a language server, as the LSP protocol assumes 0-based positions. It would be nice to switch to using 0-based indexing for better compatibility with other tools.

**Comment by pepone** [comment]
If we change this we will have to fix the MSBuild task https://github.com/icerpc/icerpc-csharp/blob/0b36976cc3acb632e21c85f2982153cc0872f1ef/tools/IceRpc.Slice.Tools/SliceCCSharpTask.cs#L134-L142

So far there would be always tools that are 1 based or 0 based, we should pick one and stick with it.

**Comment by ReeceHumphreys** [comment]
> If we change this we will have to fix the MSBuild task https://github.com/icerpc/icerpc-csharp/blob/0b36976cc3acb632e21c85f2982153cc0872f1ef/tools/IceRpc.Slice.Tools/SliceCCSharpTask.cs#L134-L142
> 
> So far there would be always tools that are 1-based or 0-based, we should pick one and stick with it.

Yes, I just think 0 based is more natural - we should do a survey of common tools and see if more are 0 based or 1 based. Generally for programming im inclined to assume things are 0 based.

**Comment by InsertCreativityHere** [comment]
### Java uses 1-based spans. Not 0 based.


At least internally, The Java compiler seems to be 1-based, not 0-based:
https://github.com/openjdk/jdk/blob/84afc28148452fb94f7d5497e512d4847e0f7a74/src/jdk.compiler/share/classes/com/sun/source/tree/LineMap.java#L46-L57

----

Experimentally, the diagnostics it prints to the command line are also 1-based.
Compiling this file (no leading space):
```java
package 0_foo;

public class MyClass {
    public static void main(String[] args) {

    }
}
```
gives:
```powershell
PS D:\Code\Workspace\temp-temp\temp-again> javac test.java
test.java:1: error: illegal underscore
package 0_foo;
         ^
1 error
```
The first line is described as `test.java:1`

**Comment by InsertCreativityHere** [comment]
## Rust Uses 1-based spans. Not 0 based.

Internally, Rust uses 1-based spans. The clearest example of that is in this `source_map` file in `compiler::rustc_span`:
https://github.com/rust-lang/rust/blob/8068e2fc9afa8c888336b12db01987be768785f9/compiler/rustc_span/src/source_map.rs#L527-L560
It states:
> Be careful because the line numbers in Loc are 1-based, so we subtract 1 to get 0-based lines.

So, the locations stored in the spans are 1-based. Here it's subtracting 1 because it's dealing with a collection of strings (and obviously, collections are 0-based).

----

Experimentally, the diagnostics it prints to the command line are also 1-based:
Compiling this file (no leading space):
```rust
use crate::bad;
// Copyright (c) ZeroC, Inc.
...
```
gives:
```powershell
error[E0432]: unresolved import `crate::bad`                                                                                                                                                                                                                                                   
 --> slicec\src\parsers\common.rs:1:5
  |
1 | use crate::bad;
  |     ^^^^^^^^^^ no `bad` in the root

Some errors have detailed explanations: E0432, E0753.
```
The first line is described as `common.rs:1`, and the column is listed as `5`, indicating columns also start at 1.

Unfortunately, I can't any direct evidence of how they're stored/processed internally.

**Comment by externl** [comment]
Unless I'm missing something the lines in a file always start at `1`. I'm actually surprised any tooling would do 0-based file spans.

**Comment by InsertCreativityHere** [comment]
Well it could be a difference in "how these values are stored internally" vs "how they're printed to the console".
Indeed, I think every printed diagnostic starts at (1,1). But internally, there's really no reason why you couldn't pick `0`.

`0` saves you a conversion when you're looking up snippets (since programmatic indices start at 0).
`1` saves you a conversion when you print line numbers to the console.

^ Purely my own speculation.

**Comment by InsertCreativityHere** [comment]
I thought about checking other languages, but I kinda forgot. Regardless, I'm going to stick with our current 1-based spans. There's precedent for it in Java and Rust, and I think it's clearer/more expected for our spans to store the information in the way it will be represented to the user (files start at line 1).

Unfortunately, this does mean we'll have to be careful about doing conversions.
And I'm going to leave this issue open as a reminder to go through and make sure we properly explain and document that our spans are 1-based everywhere, since we're probably lacking in documenting this non-obvious decision.
